debug: coredump: add new backend to store dump in flash, shell support and misc fixes

dcpleung opened the pull request:
This adds a new backend to use a flash partition to store coredump, and associated shell module to retrieve the stored coredump. 

This also fixes an issue in the Python script where a non-existent memory region would crash the script.

dcpleung committed:
debug: coredump: remove unused code

Originally there was a null backend but it was removed before
code was merged. However, some leftover code still refers to
this null backend. So remove the leftovers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug: coredump: add new backend using flash partition

This adds another simple backend to use flash partition as
backing storage.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug: coredump: add query and command APIs

This adds two new APIs to the coredump subsystem to perform
query and command. These can be used to query coredump subsys
for information, and to perform commands such as finding
out if there is a stored coredump.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: debug: add new coredump_backends

This adds a new coredump_backends test for coredump backends,
and currently tests both the logging and flash partition
backends.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug: coredump: no need to track errors at frontend

The coredump frontend (mostly) consists of wrappers of backend
functions so there is really no need to track errors at
the frontend level. Let the backends deal with their own errors
and this simplifies the code a bit.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: coredump: fix crashing on non-existent memory region

get_mem_region() may return None for a non-existent memory region
so we need to check first before looking at its properties.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
debug: coredump: add shell commands for logging backend

This adds some shell commands to the logging backend. Since
this is a simple backend, only get/clear errors are
implemented.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

nashif commented on include/debug/coredump.h:
```diff
@@ -136,6 +170,30 @@ void z_coredump_buffer_output(uint8_t *buf, size_t buflen)
  */
 
+/**
+ * @fn int z_coredump_query(enum z_coredump_query_id query_id, void *arg);
```
are those public APIs? if yes, we should not be using z_, z_ is only for private APIs used within the subsystem itself, not applicaton or user facing.

dcpleung replied:
My thinking is that these are not to be used as-is, and will always be used under another layer of subsystem (e.g. telemetry). That's why these are not public facing APIs.

nashif replied:
ok, we should remove z_ prefix. z_ is basically a replacement for plain _ and signifies a private function within a subsystem. This is a public header and the APIs have documentation and will be used to add new backends and such, so it is not private and defines an interface. Even if not application facing.

dcpleung replied:
Removed the z_ prefix in public APIs.

dcpleung committed:
debug: coredump: remove z_ prefix for stuff used outside subsys

This removes the z_ prefix those (functions, enums, etc.) that
are being used outside the coredump subsys. This aligns better
with the naming convention.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

jenmwms commented on subsys/debug/coredump/coredump_core.c:
```diff
@@ -158,3 +158,29 @@ void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr)
+		ret = backend_api->query(query_id, arg);
+	}
+
+	return ret;
```
just a thought - we could condense this, unless we see it scaling up to different conditional statements. 
`return (backend_api->query == NULL) ? (-ENOTSUP) : (backend_api->query(query_id, arg));`

nashif replied:
the code as it is is easier to read, you can add comment, if it changes, you it is easier to review in form of a patch, and above all, there is no gain in making this a one liner.

jenmwms replied:
ok, understood. it was just a thought.

jenmwms commented on subsys/debug/coredump/coredump_core.c:
```diff
@@ -158,3 +158,29 @@ void z_coredump_memory_dump(uintptr_t start_addr, uintptr_t end_addr)
+		ret = backend_api->cmd(cmd_id, arg);
+	}
+
+	return ret;
```
similar comment here about condensing this part. not a blocker.

jenmwms replied:
nvm see above reply
